extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/QuarkusRecoveryService.java

```diff
@@ -1,17 +1,54 @@
 package io.quarkus.narayana.jta.runtime;
 
+import java.util.ArrayList;
+import java.util.List;
+
+import org.jboss.tm.XAResourceRecovery;
+
 import com.arjuna.ats.jbossatx.jta.RecoveryManagerService;
 
-public class QuarkusRecoveryService {
+public class QuarkusRecoveryService extends RecoveryManagerService {
     private static RecoveryManagerService recoveryManagerService;
+    private List<XAResourceRecovery> xaResources;
+    private boolean isCreated;
 
     public static RecoveryManagerService getInstance() {
         if (recoveryManagerService == null) {
-            recoveryManagerService = new RecoveryManagerService();
+            recoveryManagerService = new QuarkusRecoveryService();
         }
         return recoveryManagerService;
     }
 
     private QuarkusRecoveryService() {
+        xaResources = new ArrayList<>();
+        isCreated = false;
+    }
+
+    @Override
+    public void addXAResourceRecovery(XAResourceRecovery xares) {
+        if (isCreated) {
+            super.addXAResourceRecovery(xares);
+        } else {
+            xaResources.add(xares);
+        }
+    }
+
+    @Override
+    public void removeXAResourceRecovery(XAResourceRecovery xares) {
+        if (isCreated) {
+            super.removeXAResourceRecovery(xares);
+        } else {
+            xaResources.remove(xares);
+        }
+    }
+
+    @Override
+    public void create() {
+        super.create();
+        isCreated = true;
+        for (XAResourceRecovery xares : xaResources) {
+            super.addXAResourceRecovery(xares);
```

Comment:
I think there is no such race condition in this case since we create and start recovery manager just after all of DataSource Synthetic beans have been initialized. Also there is a single thread when bootstrap the application at runtime.

Reply:
That is useful information, thanks for clarifying.